Fix NormalizePath trailing slash behavior

Fabianstelmach opened the pull request:
Fixes #1536

NormalizePath would append a trailing slash if and only if it would
merge multiple slashes into one. For example, the following path would
not get a trailing slash appended:

```"/sample/path"```

On the other hand, this path would get a trailing slash appended:

```"/sample//path"```

After this commit, the trailing slash is appended for both the paths.

Also, the documentation for NormalizePath is updated to inform the user of the trailing slash behavior.

Fabianstelmach committed:
Fix NormalizePath trailing slash appending

NormalizePath would append a trailing slash if and only if it would
merge multiple slashes into one. For example, the following path would
not get a trailing slash appended:

"/sample/path"

On the other hand, this path would get a trailing slash appended:

"/sample//path"

After this commit, the trailing slash is appended for both the paths.

Fabianstelmach committed:
Update NormalizePath documentation to reflect trailing slash behavior

stevemk14ebr commented:
There's no need for the let append_trailing_slash = !path.ends_with("/"); or the branching. Just move the  let original_len = path.len(); check before the append like originally discussed in the issue.

Fabianstelmach commented:
Comparing original length and the new path length is not enough in some cases. As an example, take the following path of length 5: ```/s//p``` This path would get normalized to: ```/s/p/```

But this path also has length of 5. Therefore, comparing the length of those paths is not enough to determine whether the path has changed. I do not see how this can be solved using just path length comparison, but maybe I'm missing something?

In my eyes, there are two possible approaches for this:

1) Check whether a trailing slash has been appended OR whether any slashes have been removed with the `replace_all` method.

2) Compare the old path and new path using string comparison directly.

In this PR, the first method has been implemented. If preferred, I could implement the second method instead. 

Please let me know whether either of those solutions is correct.

stevemk14ebr commented:
Ah I understand, with the case you just mentioned this seems fine.

robjtede commented:
I'd be in favor of documenting the whys and edge cases covered in code comments.

cetra3 commented:
My understanding is the intent is to see whether the path has had any changes to it by doing a length check, which fails in certain scenarios.  I think `replace_all` will return `Cow` so it should be enough to check if it's borrowed (I.e, no changes). I.e,

```rust
match self.merge_slash.replace_all(&path, "/")
  Cow::Borrowed(_var) => self.service.call(req),
  Cow::Owned(new_path) => {
  ....// we have a new path here
} 
```

You should also add tests for the example you have provided

Fabianstelmach committed:
Add normalize middleware edge case tests

Fabianstelmach commented:
> 
> 
> I'd be in favor of documenting the whys and edge cases covered in code comments.

Agreed, this is now covered in the code.



> 
> 
> My understanding is the intent is to see whether the path has had any changes to it by doing a length check, which fails in certain scenarios. I think `replace_all` will return `Cow` so it should be enough to check if it's borrowed (I.e, no changes). I.e,
> 
> ```rust
> match self.merge_slash.replace_all(&path, "/")
>   Cow::Borrowed(_var) => self.service.call(req),
>   Cow::Owned(new_path) => {
>   ....// we have a new path here
> } 
> ```

This works well for determining whether any duplicate slash has been dropped, but it does not help determine whether a trailing slash has been appended, as the latter does not use regex.

> You should also add tests for the example you have provided

Agreed, and implemented.

I have changed the implementation to use string comparison instead. The code is much simpler now. Please let me know if this solution is appropriate.

cetra3 commented:
I still can't see a test for the scenario you're trying to test against, although the implementation doesn't matter now, but I would still like to see something to guard against it if the implementation changes.

Fabianstelmach commented on src/middleware/normalize.rs:
```diff
@@ -178,11 +199,11 @@ mod tests {
 
         let srv = |req: ServiceRequest| {
-            assert_eq!(URI, req.path());
+            assert_eq!(URI.to_string() + "/", req.path());
```
Testing the case mentioned in the issue #1536, a simple path gets trailing slash appended.

Fabianstelmach commented on src/middleware/normalize.rs:
```diff
@@ -156,6 +173,10 @@ mod tests {
+
+        let req4 = TestRequest::with_uri("/v1//something").to_srv_request();
+        let res4 = normalize.call(req4).await.unwrap();
+        assert!(res4.status().is_success());
```
This test case would not work with the previous implementation using just string length. The path `/v1//something` will get normalized to `/v1/something/` and so its length won't change.

Fabianstelmach commented on src/middleware/normalize.rs:
```diff
@@ -131,6 +144,10 @@ mod tests {
+
+        let req4 = TestRequest::with_uri("/v1//something").to_request();
+        let res4 = call_service(&mut app, req4).await;
+        assert!(res4.status().is_success());
```
This test case would not work with the previous implementation using just string length. The path `/v1//something` will get normalized to `/v1/something/` and so its length won't change.

Fabianstelmach reviewed (commented):
I have updated an old test case and added a new one, based on the issue #1548 and the discussion here. I highlighted those tests in the review comments. 

Do those tests not suffice? If so, what kind of tests would you suggest to add?

cetra3 reviewed (approved):
Approved.

robjtede commented:
Just a changelog entry to add now.

robjtede committed:
Merge branch 'master' into master

robjtede committed:
Update CHANGES.md
